kimchi/src/circuits/polynomials/poseidon.rs

```diff
@@ -76,30 +87,28 @@ pub fn constraint<F: FftField + SquareRootField>() -> E<F> {
         })
         .collect();
 
-    for e in ROUND_EQUATIONS.iter() {
-        let &RoundEquation {
-            source,
-            target: (target_row, target_round),
-        } = e;
-        let sboxed: Vec<_> = round_to_cols(source)
-            .map(|i| {
-                cache.cache(
-                    E::cell(Column::Witness(i), Curr).pow(PlonkSpongeConstants15W::SPONGE_BOX),
-                )
-            })
-            .collect();
+    for (round, eq) in ROUND_EQUATIONS.iter().enumerate() {
+        // sbox
+        let mut sboxed = vec![];
+        for input_col in eq.input_cols.clone() {
```

Comment:
I find this mutability significantly less readable than the original.

Reply:
For this part, I am also inclined to think that the original approach seems more direct (...) _continue in line 102_

Reply:
my opinion is that the imperative way is more idiomatic in Rust and more readable (unless you come from a functional language). Also it most of the time takes less lines of code.